doc/developer/path-internals-daemon.rst

```diff
@@ -0,0 +1,111 @@
+PATHD Internals
+===============
+
+Architecture
+------------
+
+Overview
+........
+
+The pathd deamon manage the segment routing policies, it owns the data
+structures representing them and can load modules that manipulate them like the
+PCEP module. Its responsibility is to select a candidate path for each configured
+policies and installed it in Zebra.
+
+Zebra
+.....
+
+Zebra manage policies that are active or pending to be activated due to the next
+hop not being available yet. In zebra, policies structure and API is defined in
+`zebra_srte.[hc]`.
+
+The responsibilities of Zebra are:
+
+ - Store the policies segment list.
+ - Install the policies when their next-hop is available.
+ - Notify other daemons of the status of the policies.
+
+Adding and removing policies is done using the commands `ZEBRA_SR_POLICY_SET`
+and `ZEBRA_SR_POLICY_DELETE` as parameter of the function `zebra_send_sr_policy`
+all defined in `zclient.[hc]`.
+
+If the first segment of the policy is an unknown label, it is kept until
+notified by the mpls hooks `zebra_mpls_label_created`, and then it is installed.
+
+To get notified when a policy status changes, a client can implement the
+`sr_policy_notify_status` callback defined in `zclient.[hc]`.
+
+For encoding/decoding the various data structures used to comunicate with zebra,
+the following functions are available from `zclien.[hc]`:
+`zapi_sr_policy_encode`, `zapi_sr_policy_decode` and
+`zapi_sr_policy_notify_status_decode`.
+
+
+Pathd
+.....
+
+
+Pathd daemon manages all the possible candidate paths for the segment routing
+policies and selects the best one following the configured rules. It also
+supports loadable modules for handling dynamic candidate paths and the creation
+of new policies and candidate paths at runtime.
+
+The responsibilities of the pathd base daemon (without module) are:
+
+ - Store the policies and all the possible candidate paths for them.
+ - Select the best candidate path for each policy and send them to Zebra.
+ - Provide VTYSH configuration to set up policies and candidate paths.
+ - Provide a Northbound API to manipulate **configured** policies and candidate paths.
+ - Handle loadable modules for extending the functionality.
+ - Provide an API to the loadable module to manipulate policies and candidate paths.
+
+
+Threading Model
+---------------
+
+The daemon runs completely inside the main thread using FRR even model, there is
+no threading involved.
+
+
+Source Code
+-----------
+
+Internal Data Structures
+........................
+
+The main data structures for policies and candidate path are defined in
+`pathd.h` and implemented in `pathd.c`.
+
+When modifying these structures, either directly or through the functions
+exported by `pathd.h`, nothing should be deleted/freed right away. The deletion
```

Comment:
freed/free'd ? (not sure)

Reply:
I look it up a bit and seem to be freed...